### PR TITLE
Architecture Overhaul, and Performance Improvements

### PR DESCRIPTION
**Pull request by xadahiya**
## Summary

Addresses and closes issue #47 while introducing significant architectural enhancements and performance optimizations.

### Checklist
- [x] Ensured my branch is up-to-date with the upstream/develop branch.
- [x] Verified functionality and conducted thorough testing for Python versions 3.8.0 and above.
- [x] Ran pre-commit checks against my changes.

### Changes

Incorporates fixes from PRs #45, #50, and #53.

### New Architecture

This revamped architecture transforms the snapshotter into a more modular and highly configurable system, fostering easy customization and seamless integration. The architecture comprises three core components:

1. **Main Snapshotter Codebase**:
   - Serves as the foundational element, defining crucial interfaces and handling a diverse set of tasks, from monitoring epoch release events to task distribution and snapshot management.

2. **Configuration Files**:
   - Located in the `/config` directory, these files play a pivotal role in defining project types, specifying paths for individual compute modules, and managing various project-related settings.

3. **Compute Modules**:
   - Housed in the `snapshotter/modules` directory, these modules constitute the system's core, defining the computation logic for each project type and steering the snapshot generation process.

SnapshotterArchitecture

This architecture promotes the seamless interchange of configurations and modules. By maintaining these components in separate Git repositories, integrated through Git Submodules, adapting the system to different use cases becomes as straightforward as changing a Git branch. This approach offers unparalleled flexibility and versatility.

Learn more about using Git Submodules in the [Git Submodules Documentation](https://git-scm.com/book/en/v2/Git-Tools-Submodules).

### Other Changes and Improvements
- Added Docstrings for all major interfaces
- Introduced support for [IPFS remote pinning](https://docs.ipfs.tech/how-to/work-with-pinning-services/)
- Enhanced health check and respawning mechanisms
- Unified Phase 1 and Phase 2 architecture into a single snapshotter node
- Redesigned preloading architecture for improved performance reliability
- Moved IPFS uploads to be handled from the snapshotter node instead of Audit protocol to keep RabbitMQ messages lightweight
- Implemented bug fixes and logging improvements
- Updated README for clarity and completeness.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,37 +19,59 @@
...
+* Each data point is calculated, updated, and synchronized with other snapshotter peers participating in the network
 * synchronization of data points is defined as a function of an epoch ID(identifier) where epoch refers to an equally spaced collection of blocks on the data source blockchain (for eg, Ethereum Mainnet/Polygon Mainnet/Polygon Testnet -- Mumbai). This simplifies the building of use cases that are stateful (i.e. can be accessed according to their state at a given height of the data source chain), synchronized, and depend on reliable data. For example,
     * dashboards by offering higher-order aggregate datapoints
     * trading strategies and bots
 * a snapshotter peer can load past epochs, indexes, and aggregates from a decentralized state and have access to a rich history of data
     * all the datasets are decentralized on IPFS/Filecoin
     * the power of these decentralized storage networks can be leveraged fully by applying the [principle of composability](#aggregation-and-data-composition---snapshot-generation-of-higher-order-datapoints-on-base-snapshots)
 
+### Architecture
```
Comment (anomit) [inline README.md:53]: * We should link to the `snapshotter-computes` and `snapshotter-config` repos. 
* Either in this section, or in the case studies section, highlight the computes and configs specific to their snapshotter implementation

File: README.md [hunk 2]
```diff
@@ -431,33 +505,95 @@ time_in_seconds = 86400
...
+        "class_name": "BungeeBridgeProcessor"
+      }
+    },
+```
+Its preloader dependency is [`block_transactions`](snapshotter/utils/preloaders/tx_receipts/preloader.py) as seen in the [preloader configuration](#preloading).
 
+The snapshot builder then goes through all preloaded block transactions, filters out, and then generates relevant snapshots for wallet address that received funds from the Bungee Bridge refuel contract during that epoch.
 
+https://github.com/PowerLoom/snapshotter-computes/blob/29199feab449ad0361b5867efcaae9854992966f/bungee_bridge.py#L40-L92
 ## Find us
```
Comment (anomit) [inline README.md:597]: Let's link to our new blog at https://blog.powerloom.io/

**Review by anomit — APPROVED** [review]
LGTM! 🚀

File: README.md [hunk 3]
```diff
@@ -1,11 +1,14 @@
 ## Table of Contents
 - [Table of Contents](#table-of-contents)
 - [Overview](#overview)
+  - [Architecture](#architecture)
 - [Setup](#setup)
 - [State transitions and data composition](#state-transitions-and-data-composition)
```
Comment (anomit) [inline README.md:6]: The following sections seem to be missing:

* [Epoch processing state transitions](https://github.com/PowerLoom/pooler/tree/feat/single_snapshotter#epoch-processing-state-transitions)
* [Internal Snapshotter APIs](https://github.com/PowerLoom/pooler/tree/feat/single_snapshotter#internal-snapshotter-apis)

Comment (xadahiya) [reply README.md:6]: Added!

**Pull request merged**